### ISSUE TITLE
Standard Deviation for Evaluation Metrics

### ISSUE THREAD

**Issue by ansariyusuf** [keras-team/keras#15584]
**System information**.

TensorFlow version (you are using): TF 1.15
Are you willing to contribute it (Yes/No): yes

**Describe the feature and the current behavior/state**.

To be able to obtain the standard deviation of metrics during model training and when using model.evaluate

**Will this change the current API? How?**

I am not sure, probably it will add an additional flag to the function call.

**Who will benefit from this feature?**

This will provide more useful information, allowing everyone to understand how consistent is the model over the train/test set.

**Comment by chunduriv** [comment]
@ansariyusuf, Can you please elaborate about your Feature and please specify the Use Cases for this feature. Thanks!

**Comment by ansariyusuf** [comment]
Detailed Feature Explanation:
In Keras, if model.fit or model.evaluate is called, the metrics and loss are printed. E.g., if the model is compiled with BCE loss and metrics like dice coefficient and IoU, then these will be printed. However, there is no way of obtaining (In my current knowledge) the metrics' standard deviation. If I evaluate my models on a dataset, I get the metric value, but I don't get the standard deviation. It will be beneficial to have this built-in; it will make metrics more meaningful and quantify how consistent the model is. 

Use Case: This will benefit everyone curious to know how consistent the model is and whether metrics accurately depict the model's overall performance.

**Comment by sachinprasadhs** [comment]
Since we no longer support Tensorflow 1.x issues, could you please update your request to Tensorflow 2.x related. Thanks

**Comment by mattdangerw** [comment]
@ansariyusuf when were you thinking a metric should be sampled to calculate a standard deviation?

Just talked through this in our triage meeting: we are skeptical of per batch std dev or per epoch std dev on metrics, this would not be a generally meaningful quantity. Std dev of a metric across multiple runs would have meaning, but that could be done today without adding anything at the Keras level.

A user could do this themselves at the per epoch level using the history object returned from `model.fit`. Or at the per batch level by calling train on batch themselves which would include batch level metric values. But we don't believe this is something we should support in the Keras training APIs themselves.

**Comment by ivkireev86** [comment]
Hello, I have the same request to add standard deviation to logged metrics.
My case is about hyperparameters tuning. I have some set of hparams, and I evaluate a model with crossvalidation.
Finally I have the model quality for each fold. Currently I can log to tensorboard only mean metric. But I would like see mean with std to estimate a confidence interval.

**Comment by seanv507** [comment]
@mattdangerw 

> @ansariyusuf when were you thinking a metric should be sampled to calculate a standard deviation?
> 
> Just talked through this in our triage meeting: we are skeptical of per batch std dev or per epoch std dev on metrics, this would not be a generally meaningful quantity. Std dev of a metric across multiple runs would have meaning, but that could be done today without adding anything at the Keras level.
> 
> A user could do this themselves at the per epoch level using the history object returned from `model.fit`. Or at the per batch level by calling train on batch themselves which would include batch level metric values. But we don't believe this is something we should support in the Keras training APIs themselves.

I'm not sure I understand why you think this is not a meaningful quantity.  I think the request is addressed to the very common  'mean' metrics eg Mean square error or log loss, the idea is very simple -> the standard error of the sample mean of X is the standard deviation of the X (at sample level) /sqrt(number of samples) https://en.wikipedia.org/wiki/Standard_error

I understand this is not something that can be done for all metrics, but the common mean metrics would seem to be important enough to be supported.
